File: DTensor/Tensor-Implementations/Tests/OptimizerTests/test_drift_1024_gpu.py
#!/usr/bin/env python3
"""
test_drift_1024_gpu.py

GPU-only long-horizon drift test: 500 steps, 1024x1024 tensors.
Runs PyTorch CUDA reference for Adam and AdamW, then compares against C++ CSV.

Uses the same LCG PRNG as C++ for gradient generation.
Only 32 sampled elements (first 16 + last 16) are compared per step.

Usage:
  1. First run the C++ test:
       make run-snippet FILE=Tests/OptimizerTests/test_drift_1024_gpu.cpp
  2. Then run this script:
       cd Tests/OptimizerTests && python3 test_drift_1024_gpu.py
"""
import torch
import csv
import os
import sys

# ============================================================================
# Configuration
# ============================================================================
NUM_STEPS = 5000
ROWS = 1024
COLS = 1024
NUMEL = ROWS * COLS  # 1,048,576

# Sample indices: first 16 + last 16 (must match C++)
NUM_SAMPLES = 32
SAMPLE_INDICES = list(range(16)) + list(range(NUMEL - 16, NUMEL))


# ============================================================================
# Deterministic LCG PRNG — must match C++ exactly
# ============================================================================
def lcg_float(seed: int) -> float:
    """Returns a float in [-1.0, 1.0) given a seed. Matches C++ lcg_float()."""
    MOD = 2**64
    A = 6364136223846793005
    C = 1442695040888963407
    state = seed % MOD
    state = (A * state + C) % MOD
    state = (A * state + C) % MOD
    bits = (state >> 33) & 0x7FFFFFFF
    return (float(bits) / float(0x7FFFFFFF)) * 2.0 - 1.0


def make_gradient(step: int, device: torch.device) -> torch.Tensor:
    """Generate 1024x1024 gradient with deterministic values matching C++.
    
    Only computes the 32 sampled elements' positions fully via LCG, but we need
    ALL elements to match C++ since the optimizer processes the full tensor.
    We generate the full tensor using LCG for correctness.
    """
    # For 1M elements, pure Python LCG is slow (~2s/step).
    # Use vectorized approach: pre-compute all seeds, run LCG in numpy
    import numpy as np
    
    seeds = np.arange(NUMEL, dtype=np.uint64) + np.uint64(step * 2000000)
    
    A = np.uint64(6364136223846793005)
    C = np.uint64(1442695040888963407)
    
    state = seeds
    state = A * state + C  # numpy uint64 naturally wraps mod 2^64
    state = A * state + C
    
    bits = (state >> np.uint64(33)).astype(np.uint32) & np.uint32(0x7FFFFFFF)
    values = (bits.astype(np.float32) / np.float32(0x7FFFFFFF)) * 2.0 - 1.0
    
    grad = torch.from_numpy(values.reshape(ROWS, COLS)).to(device)
    return grad


# ============================================================================
# CSV utilities
# ============================================================================
def write_csv(path: str, steps_data: list):
    with open(path, 'w', newline='') as f:
        writer = csv.writer(f)
        header = ['step'] + [f'p{idx}' for idx in SAMPLE_INDICES]
        writer.writerow(header)
        for step, values in steps_data:
            row = [step] + [f'{v:.8f}' for v in values]
            writer.writerow(row)


def read_csv(path: str):
    data = {}
    with open(path, 'r') as f:
        reader = csv.reader(f)
        next(reader)
        for row in reader:
            step = int(row[0])
            values = [float(x) for x in row[1:]]
            data[step] = values
    return data


def sample_tensor(t: torch.Tensor) -> list:
    flat = t.detach().cpu().flatten()
    return [flat[idx].item() for idx in SAMPLE_INDICES]


# ============================================================================
# PyTorch reference runners (GPU only)
# ============================================================================
def make_initial_weights(device: torch.device) -> torch.Tensor:
    W = torch.zeros(ROWS, COLS, dtype=torch.float32, requires_grad=True, device=device)
    with torch.no_grad():
        flat = W.view(-1)
        for i in range(NUMEL):
            flat[i] = 0.1 * (i + 1) / float(NUMEL)
    return W


def run_adam(device: torch.device, csv_path: str):
    print(f"\n=== PyTorch Adam (CUDA) — {NUM_STEPS} steps, [{ROWS}x{COLS}] ===")
    W = make_initial_weights(device)
    optimizer = torch.optim.Adam([W], lr=0.001, betas=(0.9, 0.999), eps=1e-8, weight_decay=0.01)

    steps_data = [(0, sample_tensor(W))]
    for step in range(1, NUM_STEPS + 1):
        optimizer.zero_grad()
        W.grad = make_gradient(step, device)
        optimizer.step()
        steps_data.append((step, sample_tensor(W)))
        if step % 100 == 0:
            print(f"  Step {step}/{NUM_STEPS}")

    write_csv(csv_path, steps_data)
    print(f"  Wrote {csv_path}")


def run_adamw(device: torch.device, csv_path: str):
    print(f"\n=== PyTorch AdamW (CUDA) — {NUM_STEPS} steps, [{ROWS}x{COLS}] ===")
    W = make_initial_weights(device)
    optimizer = torch.optim.AdamW([W], lr=0.001, betas=(0.9, 0.999), eps=1e-8, weight_decay=0.01)

    steps_data = [(0, sample_tensor(W))]
    for step in range(1, NUM_STEPS + 1):
        optimizer.zero_grad()
        W.grad = make_gradient(step, device)
        optimizer.step()
        steps_data.append((step, sample_tensor(W)))
        if step % 100 == 0:
            print(f"  Step {step}/{NUM_STEPS}")

    write_csv(csv_path, steps_data)
    print(f"  Wrote {csv_path}")


# ============================================================================
# Comparison
# ============================================================================
def compare_csvs(own_path: str, pytorch_path: str, name: str, tolerance: float = 1e-5):
    if not os.path.exists(own_path):
        print(f"  [SKIP] {name}: C++ CSV not found at {own_path}")
        print(f"         Run: make run-snippet FILE=Tests/OptimizerTests/test_drift_1024_gpu.cpp")
        return None

    own_data = read_csv(own_path)
    pytorch_data = read_csv(pytorch_path)

    max_dev = 0.0
    worst_step = 0
    worst_elem_idx = 0
    num_steps_checked = 0

    for step in sorted(pytorch_data.keys()):
        if step not in own_data:
            print(f"  [FAIL] {name}: Step {step} missing from C++ output")
            return False

        own_vals = own_data[step]
        pt_vals = pytorch_data[step]

        for i in range(len(pt_vals)):
            dev = abs(own_vals[i] - pt_vals[i])
            if dev > max_dev:
                max_dev = dev
                worst_step = step
                worst_elem_idx = i
        num_steps_checked += 1

    passed = max_dev < tolerance

    print(f"\n  --- {name} ---")
    print(f"  Tensor shape    : [{ROWS}x{COLS}] ({NUMEL:,} elements)")
    print(f"  Elements sampled: {NUM_SAMPLES}")
    print(f"  Steps compared  : {num_steps_checked}")
    print(f"  Max deviation   : {max_dev:.2e}")
    print(f"  Worst at        : step={worst_step}, element=p{SAMPLE_INDICES[worst_elem_idx]}")
    if passed:
        print(f"  Result          : [PASS] (tolerance={tolerance:.0e})")
    else:
        own_val = own_data[worst_step][worst_elem_idx]
        pt_val = pytorch_data[worst_step][worst_elem_idx]
        print(f"  OwnTensor value : {own_val:.8f}")
        print(f"  PyTorch value   : {pt_val:.8f}")
        print(f"  Result          : [FAIL] deviation {max_dev:.2e} > tolerance {tolerance:.0e}")

    return passed


# ============================================================================
# Main
# ============================================================================
def main():
    print("=" * 70)
    print("  GPU Drift Test — 1024x1024 Tensors, 500 Steps")
    print("  PyTorch Reference + Comparison")
    print("=" * 70)

    if not torch.cuda.is_available():
        print("ERROR: CUDA not available. This test requires a GPU.")
        sys.exit(1)

    device = torch.device("cuda")
    print(f"CUDA: {torch.cuda.get_device_name(0)}")

    # Run PyTorch references
    print("\n" + "-" * 40)
    print("Running PyTorch references...")
    print("-" * 40)

    run_adam(device, "drift_1024_adam_gpu_pytorch.csv")
    run_adamw(device, "drift_1024_adamw_gpu_pytorch.csv")

    # Compare
    print("\n" + "=" * 70)
    print("  Comparing OwnTensor vs PyTorch (GPU, 1024x1024)")
    print("=" * 70)

    results = []
    results.append(compare_csvs("drift_1024_adam_gpu.csv", "drift_1024_adam_gpu_pytorch.csv", "Adam (GPU)"))
    results.append(compare_csvs("drift_1024_adamw_gpu.csv", "drift_1024_adamw_gpu_pytorch.csv", "AdamW (GPU)"))

    # Summary
    print("\n" + "=" * 70)
    valid = [r for r in results if r is not None]
    passed = sum(1 for r in valid if r)
    total = len(valid)
    skipped = len(results) - total

    print(f"  Overall: {passed}/{total} PASSED", end="")
    if skipped > 0:
        print(f", {skipped} SKIPPED", end="")
    print()

    if passed == total and total > 0:
        print("All GPU 1024x1024 drift tests PASSED!")
    else:
        print("Some tests FAILED or were skipped.")
    print("=" * 70)

    sys.exit(0 if passed == total else 1)


if __name__ == "__main__":
    main()

File: DTensor/Tensor-Implementations/Tests/OptimizerTests/test_drift_1024_gpu.cpp
/**
 * @file test_drift_1024_gpu.cpp
 * @brief GPU-only long-horizon drift test: 500 steps, 1024x1024 tensors, non-uniform gradients.
 *
 * Tests Adam and AdamW on GPU (foreach path) over 500 steps with 1M-element tensors.
 * Uses a simple LCG PRNG for gradient generation, matching the Python companion.
 * Writes sampled parameter values (first 16 + last 16 elements) to CSV.
 *
 * Build & run:
 *   make run-snippet FILE=Tests/OptimizerTests/test_drift_1024_gpu.cpp
 *
 * Then compare:
 *   cd Tests/OptimizerTests && python3 test_drift_1024_gpu.py
 */
#include <iostream>
#include <iomanip>
#include <fstream>
#include <cmath>
#include <vector>
#include <string>
#include <array>

#include "TensorLib.h"

using namespace OwnTensor;

// ============================================================================
// Configuration
// ============================================================================
static const int NUM_STEPS = 5000;
static const int ROWS = 1024;
static const int COLS = 1024;
static const int NUMEL = ROWS * COLS;  // 1,048,576

// Sample indices: first 16 + last 16 elements for CSV output
static const int NUM_SAMPLES = 32;

static std::array<int, NUM_SAMPLES> get_sample_indices() {
    std::array<int, NUM_SAMPLES> idx;
    for (int i = 0; i < 16; ++i) {
        idx[i] = i;                         // first 16: 0..15
        idx[i + 16] = NUMEL - 16 + i;      // last 16
    }
    return idx;
}

// ============================================================================
// Deterministic LCG PRNG — must match Python exactly
// ============================================================================
static float lcg_float(uint64_t seed) {
    uint64_t state = seed;
    state = (6364136223846793005ULL * state + 1442695040888963407ULL);
    state = (6364136223846793005ULL * state + 1442695040888963407ULL);
    uint32_t bits = static_cast<uint32_t>(state >> 33);
    return (static_cast<float>(bits) / static_cast<float>(0x7FFFFFFFU)) * 2.0f - 1.0f;
}


// Generate a 1024x1024 gradient tensor on CPU, then move to GPU
static Tensor make_gradient(int step) {
    TensorOptions opts = TensorOptions().with_dtype(Dtype::Float32);
    Tensor grad = Tensor::zeros(Shape{{ROWS, COLS}}, opts);
    float* data = grad.data<float>();
    for (int i = 0; i < NUMEL; ++i) {
        uint64_t seed = static_cast<uint64_t>(step) * 2000000ULL + static_cast<uint64_t>(i);
        data[i] = lcg_float(seed);
    }
    return grad.to_cuda(0);
}


// ============================================================================
// CSV utilities — only write sampled elements
// ============================================================================
static void write_csv_header(std::ofstream& f, const std::array<int, NUM_SAMPLES>& indices) {
    f << "step";
    for (int idx : indices) {
        f << ",p" << idx;
    }
    f << "\n";
}

static void write_csv_row(std::ofstream& f, int step, const Tensor& t,
                           const std::array<int, NUM_SAMPLES>& indices) {
    Tensor tc = t.to_cpu();
    const float* d = tc.data<float>();
    f << step;
    for (int idx : indices) {
        f << "," << std::fixed << std::setprecision(8) << d[idx];
    }
    f << "\n";
}


// ============================================================================
// Initial weights
// ============================================================================
static Tensor make_initial_weights() {
    TensorOptions opts = TensorOptions()
        .with_dtype(Dtype::Float32)
        .with_req_grad(true);
    Tensor W = Tensor::zeros(Shape{{ROWS, COLS}}, opts);
    float* data = W.data<float>();
    for (int i = 0; i < NUMEL; ++i) {
        data[i] = 0.1f * static_cast<float>(i + 1) / static_cast<float>(NUMEL);
    }
    W.set_requires_grad(true);
    W = W.to_cuda(0);
    W.set_requires_grad(true);
    return W;
}


// ============================================================================
// Test runners
// ============================================================================
void test_adam_gpu(const std::string& csv_path) {
    std::cout << "\n=== Adam (GPU) — " << NUM_STEPS
              << " steps, [" << ROWS << "x" << COLS << "] ===" << std::endl;

    Tensor W = make_initial_weights();
    auto indices = get_sample_indices();

    float lr = 0.001f, beta1 = 0.9f, beta2 = 0.999f, eps = 1e-8f, wd = 0.01f;
    nn::Adam adam({W}, lr, beta1, beta2, eps, wd);

    std::ofstream csv(csv_path);
    write_csv_header(csv, indices);
    write_csv_row(csv, 0, W, indices);

    for (int step = 1; step <= NUM_STEPS; ++step) {
        Tensor grad = make_gradient(step);
        W.set_grad(grad);
        adam.step();
        write_csv_row(csv, step, W, indices);
    }
    csv.close();
    std::cout << "  Wrote " << csv_path << std::endl;
}


void test_adamw_gpu(const std::string& csv_path) {
    std::cout << "\n=== AdamW (GPU) — " << NUM_STEPS
              << " steps, [" << ROWS << "x" << COLS << "] ===" << std::endl;

    Tensor W = make_initial_weights();
    auto indices = get_sample_indices();

    float lr = 0.001f, beta1 = 0.9f, beta2 = 0.999f, eps = 1e-8f, wd = 0.01f;
    nn::AdamW adamw({W}, lr, beta1, beta2, eps, wd);

    std::ofstream csv(csv_path);
    write_csv_header(csv, indices);
    write_csv_row(csv, 0, W, indices);

    for (int step = 1; step <= NUM_STEPS; ++step) {
        Tensor grad = make_gradient(step);
        W.set_grad(grad);
        adamw.step();
        write_csv_row(csv, step, W, indices);
    }
    csv.close();
    std::cout << "  Wrote " << csv_path << std::endl;
}


int main() {
    std::string base = "Tests/OptimizerTests/";

    std::cout << "================================================================" << std::endl;
    std::cout << "  GPU Drift Test — 500 steps, 1024x1024 tensors" << std::endl;
    std::cout << "  Non-uniform gradients, sampling 32 elements for CSV" << std::endl;
    std::cout << "================================================================" << std::endl;

    test_adam_gpu(base + "drift_1024_adam_gpu.csv");
    test_adamw_gpu(base + "drift_1024_adamw_gpu.csv");

    std::cout << "\n================================================================" << std::endl;
    std::cout << "  Done. Now run:" << std::endl;
    std::cout << "    cd Tests/OptimizerTests" << std::endl;
    std::cout << "    python3 test_drift_1024_gpu.py" << std::endl;
    std::cout << "================================================================" << std::endl;

    return 0;
}
